Merge failed-tests-printer and enable in CMake

tautschnig opened the pull request.

tautschnig committed:
Move implementation of failed-tests-printer.pl into test.pl

Added a new option "-p" to test.pl to request printing failure logs instead of
doing so via an external script. This simplifies various Makefiles and enables
use in cmake builds.

tautschnig committed:
Create goto-gcc symlink in cmake builds and enable goto-gcc tests

Fixes: #1500

reuk commented on regression/CMakeLists.txt:
```diff
@@ -34,6 +34,9 @@ add_subdirectory(goto-cc-goto-analyzer)
 add_subdirectory(goto-instrument)
 add_subdirectory(goto-instrument-typedef)
+if(NOT WIN32)
+  add_subdirectory(goto-gcc)
```
This will fail, as there is no CMakeLists.txt in `goto-gcc`. If you add a file with the following contents, it should work though:
```
add_test_pl_tests(
    "$<TARGET_FILE:goto-gcc>"
)
```

tautschnig replied:
Thank you for the guidance! I've added this in 4a5dff1 and will squash once CI succeeds.

reuk replied:
Think I got this wrong - if goto-gcc isn't a target, then we can't find its path in this way. As an alternative, I would hard-code the goto-gcc symlink location as `${CMAKE_BINARY_DIR}/goto-gcc`, and then use that path as the argument to `add_test_pl_tests`

tautschnig replied:
I've now gone for  cb9f586 - I feels like a horrible hack, but I haven't managed to figure out a way of marking a symlink as "executable". Maybe your hard-coded approach is the better option?

reuk reviewed (approved):
LGTM, but I think the CMakeLists for `goto-gcc` is missing. I've approved as this is a reasonably small change which won't require re-review.

tautschnig commented:
For some reason this broke AppVeyor. I'll experiment with reverting some changes; please bear with me (and don't make any attempt to merge).

tautschnig committed:
test.pl: Use native perl instead of "cat" to print log file

tautschnig committed:
Do not use shell built-ins

As we no longer use sequences of commands in the Makefile, AppVeyor's make does
not actually invoke a shell to call test.pl. Hence invoking "echo" from within
test.pl fails.

From the `make` documentation: "When it is time to execute commands to update a
target, they are executed by making a new subshell for each line. (In practice,
make may take shortcuts that do not affect the results.)" It seems these
"shortcuts" are being taken by AppVeyor's make.

tautschnig committed:
Use single quotes for Windows compatibility

tautschnig commented:
@thk123 Could I please get a review of this? It turned out to require a far bigger change than initially expected as make on AppVeyor has some interesting behaviour.

thk123 commented on regression/test.pl:
```diff
@@ -352,4 +355,38 @@ ($)
+        close FH;
+        print "\n\nFailed $descriptor_file lines:\n";
+      }
+      print "$_\n";
```
My perl isn't ~very good~ existent, but it might be more readable for non-perl users to `chomp $line;` and `print $line`. Also how come it isn't inside the `if(0 == $printed_this_test)`. (I realise this is code moved from `failed-tests-printer.pl` so feel free to say this is out of scope).

thk123 commented on src/goto-cc/CMakeLists.txt:
```diff
@@ -29,4 +29,14 @@ target_link_libraries(goto-cc goto-cc-lib)
+  # regression/goto-gcc requires an _executable_ target; this cruel hack will
+  # first build the executable and then replace it by a symbolic link so as to
+  # save some disk space
+  add_executable(goto-gcc goto_cc_main.cpp)
```
I suppose it would be good for @reuk to have a look at this work around. 

To check I understand, it is compiling `goto_cc_main.cpp` -> `goto-gcc` then deleting it by replacing it with a symlink to `goto-cc`? 

How come this is only necessary on Linux. Or rather: why can't we do `set_target_properties(goto-cc PROPERTIES OUTPUT_NAME goto-gcc)`. I realise you've spent quite a few cycles iterating to get something working, so it would be good to expand the comment to explain why it is necessary.

reuk replied:
Yes, I'd also like to know *why* regression/goto-gcc requires an executable target. Does `TARGET_FILE:goto-gcc` not work otherwise?

reuk replied:
Another option is to set an `alternative path/name` property on `goto-cc` (set to the path of `goto-gcc` and then to query it using `$<TARGET_PROPERTY:tgt,prop>` in regression/goto-gcc

tautschnig replied:
@reuk here's what I get:
```
CMake Error at regression/CMakeLists.txt:7 (add_test):
  Error evaluating generator expression:

    $<TARGET_FILE:goto-gcc>

  Target "goto-gcc" is not an executable or library.
```
Looking at CMake's documentation, this is maybe not surprising:
"$<TARGET_FILE:tgt>
    Full path to main file (.exe, .so.1.2, .a) where tgt is the name of a target."

I'm now trying some string rewriting, which I'd say is a variant on the properties approach? Comments solicited!

thk123 commented on regression/cbmc-java/swap2/test.desc:
```diff
@@ -1,6 +1,6 @@
 CORE
 org/springframework/build/gradle/MergePlugin$1$_execute_closure1$_closure2.class
---function "org.springframework.build.gradle.MergePlugin\$1\$_execute_closure1\$_closure2.\$getCallSiteArray:()[Lorg/codehaus/groovy/runtime/callsite/CallSite;"
+--function 'org.springframework.build.gradle.MergePlugin$1$_execute_closure1$_closure2.$getCallSiteArray:()[Lorg/codehaus/groovy/runtime/callsite/CallSite;'
```
I'm in relative awe at you tracking this down btw, glad we found a combination of escaped characters and quotes that is cross platform compatible...

thk123 reviewed (approved):
Everything looks reasonable and of course most importantly CI does indeed pass.

tautschnig committed:
Improve readability of code imported from failed-tests-printer.pl

tautschnig commented:
@reuk @thk123 would you mind taking another look (despite your approvals!)?

reuk commented on regression/goto-gcc/CMakeLists.txt:
```diff
@@ -0,0 +1,4 @@
+string(REGEX REPLACE "goto-cc$" "goto-gcc"
+  goto_gcc_path "$<TARGET_FILE:goto-cc>")
```
Seems reasonable. Consider adding a comment explaining why this is necessary, though.

tautschnig replied:
Thanks, done - hoping it's clear/useful?

thk123 replied:
I think so :+1:

thk123 reviewed (approved):
Seems reasonable - thanks for the improvements to the perl script.

tautschnig commented:
@reuk @thk123 Finally got it working with not too much of a hack - thanks a lot for all the guidance provided!
